File: mobile/ios/Shared/Sources/NativeVoiceProfiles.swift
import Foundation

private struct NativeBuiltinProfile: Decodable {
    struct TransformConfig: Decodable {
        let enabled: Bool
        let mode: String
        let requestFormat: String
        let customPrompt: String
        let translationSourceLanguage: String
        let translationSourceCode: String
        let translationTargetLanguage: String
        let translationTargetCode: String
        let translationExtraInstructions: String

        enum CodingKeys: String, CodingKey {
            case enabled
            case mode
            case requestFormat = "request_format"
            case customPrompt = "custom_prompt"
            case translationSourceLanguage = "translation_source_language"
            case translationSourceCode = "translation_source_code"
            case translationTargetLanguage = "translation_target_language"
            case translationTargetCode = "translation_target_code"
            case translationExtraInstructions = "translation_extra_instructions"
        }
    }

    let presetKind: String
    let id: String
    let displayName: String
    let languageTag: String
    let autoCommit: Bool
    let commitSuffixMode: String
    let transform: TransformConfig

    enum CodingKeys: String, CodingKey {
        case presetKind = "preset_kind"
        case id
        case displayName = "display_name"
        case languageTag = "language_tag"
        case autoCommit = "auto_commit"
        case commitSuffixMode = "commit_suffix_mode"
        case transform
    }
}

enum NativeVoiceProfiles {
    static func loadBuiltIns() -> [VoiceProfile]? {
        guard let raw = NativeProfileBridge.builtinProfilesJSON(),
              let data = raw.data(using: .utf8) else {
            return nil
        }

        let decoded = try? JSONDecoder().decode([NativeBuiltinProfile].self, from: data)
        return decoded?.compactMap(makeProfile)
    }

    private static func makeProfile(from native: NativeBuiltinProfile) -> VoiceProfile? {
        let mode: VoiceCommitMode
        switch native.presetKind {
        case "dictation":
            mode = .dictation
        case "chat":
            mode = .chat
        case "review":
            mode = .review
        case "translate_zh_en":
            mode = .translateChineseToEnglish
        default:
            return nil
        }

        return VoiceProfile(
            id: native.id,
            title: native.displayName,
            mode: mode,
            languageTag: native.languageTag.isEmpty ? nil : native.languageTag,
            transform: VoiceTransformConfig(
                enabled: native.transform.enabled,
                mode: VoiceTransformMode(rawValue: native.transform.mode.lowercased()) ?? .cleanup,
                requestFormat: VoiceRefineRequestFormat(rawValue: native.transform.requestFormat.lowercased()) ?? .systemAndUser,
                customPrompt: native.transform.customPrompt,
                translationSourceLanguage: native.transform.translationSourceLanguage,
                translationSourceCode: native.transform.translationSourceCode,
                translationTargetLanguage: native.transform.translationTargetLanguage,
                translationTargetCode: native.transform.translationTargetCode,
                translationExtraInstructions: native.transform.translationExtraInstructions
            )
        )
    }
}

File: mobile/android/speechcore/src/main/kotlin/com/shinsoku/mobile/speechcore/NativeVoiceProfiles.kt
package com.shinsoku.mobile.speechcore

import org.json.JSONArray

object NativeVoiceProfiles {
    init {
        runCatching { System.loadLibrary("shinsoku_nativecore") }
    }

    fun loadBuiltIns(): List<VoiceInputProfile>? {
        val raw = runCatching { builtinProfilesJsonNative() }.getOrNull().orEmpty()
        if (raw.isBlank()) {
            return null
        }

        return runCatching {
            val json = JSONArray(raw)
            buildList {
                for (index in 0 until json.length()) {
                    val item = json.optJSONObject(index) ?: continue
                    val transformJson = item.optJSONObject("transform")
                    add(
                        VoiceInputProfile(
                            id = item.optString("id"),
                            displayName = item.optString("display_name"),
                            summary = item.optString("summary"),
                            nativeBehaviorSummary = item.optString("behavior_summary"),
                            languageTag = item.optString("language_tag").ifBlank { null },
                            autoCommit = item.optBoolean("auto_commit", true),
                            commitSuffixMode = runCatching {
                                CommitSuffixMode.valueOf(item.optString("commit_suffix_mode"))
                            }.getOrDefault(CommitSuffixMode.Space),
                            transform = VoiceTransformConfig(
                                enabled = transformJson?.optBoolean("enabled", false) ?: false,
                                mode = runCatching {
                                    VoiceTransformMode.valueOf(transformJson?.optString("mode").orEmpty())
                                }.getOrDefault(VoiceTransformMode.Cleanup),
                                requestFormat = runCatching {
                                    VoiceRefineRequestFormat.valueOf(transformJson?.optString("request_format").orEmpty())
                                }.getOrDefault(VoiceRefineRequestFormat.SystemAndUser),
                                customPrompt = transformJson?.optString("custom_prompt").orEmpty(),
                                translationSourceLanguage = transformJson?.optString("translation_source_language").orEmpty().ifBlank { "Chinese" },
                                translationSourceCode = transformJson?.optString("translation_source_code").orEmpty().ifBlank { "zh" },
                                translationTargetLanguage = transformJson?.optString("translation_target_language").orEmpty().ifBlank { "English" },
                                translationTargetCode = transformJson?.optString("translation_target_code").orEmpty().ifBlank { "en" },
                                translationExtraInstructions = transformJson?.optString("translation_extra_instructions").orEmpty(),
                            ),
                        ),
                    )
                }
            }
        }.getOrNull()
    }

    fun identifyBuiltIn(profile: VoiceInputProfile): String? {
        return runCatching {
            identifyBuiltInProfileIdNative(
                autoCommit = profile.autoCommit,
                commitSuffixMode = profile.commitSuffixMode.name,
                languageTag = profile.languageTag.orEmpty(),
                transformEnabled = profile.transform.enabled,
                transformMode = profile.transform.mode.name,
                requestFormat = profile.transform.requestFormat.name,
                customPrompt = profile.transform.customPrompt,
                translationSourceLanguage = profile.transform.translationSourceLanguage,
                translationSourceCode = profile.transform.translationSourceCode,
                translationTargetLanguage = profile.transform.translationTargetLanguage,
                translationTargetCode = profile.transform.translationTargetCode,
                translationExtraInstructions = profile.transform.translationExtraInstructions,
            )
        }.getOrNull()?.ifBlank { null }
    }

    fun describeBehavior(profile: VoiceInputProfile): String {
        return runCatching {
            describeProfileBehaviorNative(
                autoCommit = profile.autoCommit,
                commitSuffixMode = profile.commitSuffixMode.name,
            )
        }.getOrNull()?.ifBlank { null } ?: fallbackBehaviorSummary(profile)
    }

    private fun fallbackBehaviorSummary(profile: VoiceInputProfile): String {
        val commitDescription = if (profile.autoCommit) "Auto-insert on" else "Review before insert"
        return "$commitDescription · ${profile.commitSuffixMode.title}"
    }

    private external fun builtinProfilesJsonNative(): String
    private external fun identifyBuiltInProfileIdNative(
        autoCommit: Boolean,
        commitSuffixMode: String,
        languageTag: String,
        transformEnabled: Boolean,
        transformMode: String,
        requestFormat: String,
        customPrompt: String,
        translationSourceLanguage: String,
        translationSourceCode: String,
        translationTargetLanguage: String,
        translationTargetCode: String,
        translationExtraInstructions: String,
    ): String
    private external fun describeProfileBehaviorNative(
        autoCommit: Boolean,
        commitSuffixMode: String,
    ): String
}
